ios/Runner/AppAuthHandler.swift

```diff
@@ -0,0 +1,304 @@
+    
+    let APP_AUTH_SESSION_CHANNEL = "insporation/appauth_authorization_events"
+    // A single slash between shema and path is recommended for iOS
+    let APP_AUTH_REDIRECT_URI =  URL.init(string: "eu.jhass.insporation:/callback/")!
+    
+    let controller : UIViewController
+    var registration : Registration?
+    let authMethodChannel : FlutterMethodChannel
+    var currentSession : Session?
+    
+    private var completionHandler : [(_ tokens:Tokens) -> Void] = []
+    private var errorHandler : [(_ errorMessage:String) -> Void] = []
+    
+    init(methodChannel: FlutterMethodChannel, controller: UIViewController) {
+        self.authMethodChannel = methodChannel
+        self.controller = controller
+        os_log("Init App Auth Handler")
+    }
+    
+    private func invokeCompletionHandler(tokens: Tokens) {
+        while !self.completionHandler.isEmpty {
+            self.completionHandler.removeFirst()(tokens)
+        }
+    }
+    
+    private func invokeErrorHandler(errorMessage : String) {
+        while !self.errorHandler.isEmpty {
+            self.errorHandler.removeFirst()(errorMessage)
+        }
+    }
+    
+    /**
+     Gets a token from auth
+     */
+    func getAccessTokens(_ session : Session, completionHandler : @escaping (Tokens) -> Void, errorHandler:@escaping (String) -> Void) {
+        self.currentSession = session
+        
+        self.completionHandler.append(completionHandler)
+        self.errorHandler.append(errorHandler)
+        
+        let state = session.state // Provided session has a state in textform, need to get a valid state from userSettings
+        
+        // A incomming session could be searche internally for existing tokens
+        
+        if (state == nil) {
+            os_log("Provided session for fetching access token has no authorization, launching authorization process")
+            authorizeUser()
+        } else {
+            
+            os_log("State was set previously, refreshing token from State..")
+            
+            // A textual state was set, recover laste stored State Obejct
+            recoverToken()
+        }
+    }
+    
+    func recoverToken() {
+        guard let session = self.currentSession else {
+            os_log("Error in getting last session")
+            self.errorHandler.first?("Error in getting last session")
+            return
+        }
+        
+        if let authState = loadState(forUserId: session.userId) {
+            // State exists, perform refreshing
+            
+            if (!authState.isAuthorized) {
+                authorizeUser()
+                return
+            }
+            
+            guard let tokenResponse = authState.lastTokenResponse else {
+                self.invokeErrorHandler(errorMessage: "No token received")
+                return
+            }
+            let tokens = Tokens(accessToken: tokenResponse.accessToken!, idToken: tokenResponse.idToken!)
```

Comment:
The access token is not valid forever and the Dart side of things does not implement token refresh itself. In the Android bridge the aforementioned `AuthState` object provides means to validate whether a given token is still valid and if not to refresh it via the refresh token: 

https://github.com/openid/AppAuth-Android/blob/master/library/java/net/openid/appauth/AuthState.java#L453

https://github.com/jhass/insporation/blob/main/android/app/src/main/kotlin/jhass/eu/insporation/AppAuthHandler.kt#L110

I  believe the equivalent in the iOS SDK is https://github.com/openid/AppAuth-iOS/blob/e191b1beadf3041259652e6e464eaceff16e5fbb/Source/AppAuthCore/OIDAuthState.h#L214-L219 ?

I couldn't spot similar serialization/deserialization capabilities as implemented for Android on first sight, but to the Dart side all the `state` things are just opaque strings :)

Reply:
Will refresh the token before using it. 
The state itself (or registration) can not be serialized on the dart side. The Objc - AuthState uses the binary NSData object to serialize and deserialize these things, which are not convertible to strings.

Reply:
Well, we could always base64 encode it....